torch/csrc/jit/codegen/cuda/transform_replay.cpp

```diff
@@ -687,134 +687,430 @@ std::deque<TensorView*> producersFor(TensorView* tv) {
+// tensor than r.
+bool operator>(const NextHopInfo& l, const NextHopInfo& r) {
+  return r < l;
+}
 
-  std::unordered_set<Val*> dep_vals_set{dep_vals.begin(), dep_vals.end()};
+// l == r means it is hard to tell which one of then contains more information
+bool operator==(const NextHopInfo& l, const NextHopInfo& r) {
+  return !(r < l) && !(l < r);
+}
 
-  auto n_transformed_root_dims = std::count_if(
-      producer_root.begin(),
-      producer_root.end(),
-      [&dep_vals_set](IterDomain* root_id) {
-        return dep_vals_set.find(root_id) != dep_vals_set.end();
-      });
+std::vector<RootIDInfo> getStartingRootIDInfo(TensorView* tv) {
+  std::vector<RootIDInfo> result;
+  auto root_domain = tv->getRootDomain();
+  result.reserve(root_domain.size());
+  for (auto id : root_domain) {
+    result.emplace_back(id);
+  }
+  return result;
+}
 
-  if (replayed_pos.find(producer_tv) != replayed_pos.end()) {
-    if (n_transformed_root_dims < n_replayed_root_dims.at(producer_tv) ||
-        (n_transformed_root_dims == n_replayed_root_dims.at(producer_tv) &&
-         replayed_producer.second <= replayed_pos.at(producer_tv))) {
-      return false; // NOLINT(clang-analyzer-cplusplus.NewDeleteLeaks)
+// Infer the compute-at position from the information of preserved reference
+// root ID.
+//
+// TODO:
+// I think I need to modify TransformReplay to add a new interface to specify
+// the root domains, instead of a position in the leaf domain. With the new
+// interface, this function will not be needed.
+size_t getReplayPos(const NextHopInfo& next_hop) {
+  auto& root_id_info = next_hop.root_id_info_from;
+  auto from_tv = next_hop.from;
+  // Flatten `root_id_info_from` to get the list of ids in the `from_tv`'s
+  // root/rfactor domain that contains information about the reference tensor.
+  std::unordered_set<IterDomain*> from_ids;
+  from_ids.reserve(root_id_info.size());
+  for (auto info : root_id_info) {
+    for (auto id : info.mapped_ids) {
+      from_ids.insert(id);
     }
   }
-
-  producer_tv->setDomain(replayed_producer.first);
-  replayed_pos[producer_tv] = replayed_producer.second;
-  n_replayed_root_dims[producer_tv] = n_transformed_root_dims;
-
-  return true;
+  // Get leaf IDs that contain information of `from_ids`
+  std::unordered_set<IterDomain*> relevant_leaves;
+  std::vector<IterDomain*> to_visit(from_ids.begin(), from_ids.end());
+  while (!to_visit.empty()) {
+    auto front = to_visit.back();
+    to_visit.pop_back();
+    if (front->uses().empty()) {
+      relevant_leaves.emplace(front);
+    } else {
+      for (auto def : front->uses()) {
+        auto outs = ir_utils::filterByType<IterDomain>(def->outputs());
+        to_visit.insert(to_visit.end(), outs.begin(), outs.end());
+      }
+    }
+  }
+  // Find the pos where all leaf IDs at <= pos contains
+  // information about the starting root domain
+  //
+  // TODO: should I change to the following behavior?
```

Comment:
Can you be more specific?

Reply:
I added a few more lines in the comment below

Reply:
Can you create a test exhibiting this behavior?

Reply:
Looks like this will never happen? I added a new test at https://github.com/csarofeen/pytorch/pull/1763/commits/5cc10c872209eadf052ea76b1fb5f88f73a3cbe3, but looks like this is not triggered. I think this is because, if `from` is the reference tensor, then we always have full information. If not, then `from` must come from a replay, then the axes containing reference tensor information will always be put to the front. So this case will never happen?

I will merge this PR for now, and will write a followup PR for `TransformReplay` starting from specified root domain. Once that is done, then this issue will not be relevant. Feel free to leave more comment, and I will resolve in my followup PR.

Reply:
For the meantime, we may want to assert that no remaining domains are included in `relevant_leaves`.

Reply:
> I'm not sure we can rewrite `TransformReplay` to use root domain positions rather than leaf domain positions. In other words, `computeAt` currently is specified with leaf positions. Can we change that to use root positions?

I think the interface and outcome of `computeAt` should still use the leaf positions. But during propagation, since we are saving information about root/rfactor domains, I think it makes sense to change the interface (or at least add an additional interface) to specify root/rfactor domains? I am not sure how doable this is. Need to dig into the code to see.

> For the meantime, we may want to assert that no remaining domains are included in `relevant_leaves`.

I think it makes sense to add an assert. If we decide that we should not change `TransformReplay` to specify the root domain, then I will add assert in a separate PR.

Reply:
We should talk more about this, root domains is a good mechanism to understand "how replayed" one replay is versus another. I'm skeptical propagating based on root domains is a good idea.